"""Fixed device inputs and captured metadata for the hybrid Qwen decoder."""

from __future__ import annotations

from dataclasses import dataclass

import torch
import triton
import triton.language as tl

from flare.components.capacity import BATCH_CAPACITIES
from flare.components.recurrent import DeltaMetadata
from flare.components.token_state import PreparedTokenState
from flare.preparation.triton import prepare_triton

from .attention import QSAMetadata
from .ple import PLEMetadata
from .sampling import DeviceRequestState
from .state import HybridStatePool


def compile_kernel(kernel, args, grid):
    return prepare_triton(kernel, args, grid, num_warps=4)


@triton.jit
def _validate(Slots, Epochs, Live, Lengths, Generation, Status, Error, Pending, Consumed,
              ROWS: tl.constexpr, TOKENS: tl.constexpr, CONTEXT: tl.constexpr,
              CONSUMED_STRIDE: tl.constexpr, PREFILL: tl.constexpr, SLOTS: tl.constexpr = 4):
    slot = tl.program_id(0)
    rows = tl.arange(0, SLOTS)
    match = (rows < ROWS) & (rows < tl.load(Live))
    match &= tl.load(Slots + rows, rows < ROWS, other=-1) == slot
    match &= tl.load(Epochs + rows, rows < ROWS, other=-1) == tl.load(Generation + slot)
    count = tl.sum(match.to(tl.int32), 0)
    if (count > 0) & (tl.load(Status + slot) == 1):
        row = tl.max(tl.where(match, rows, 0), 0)
        length = tl.load(Lengths + row) if PREFILL else 1
        consumed = tl.load(Consumed + slot * CONSUMED_STRIDE)
        valid = (count == 1) & (tl.load(Live) <= ROWS) & (length > 0) & (length <= TOKENS)
        valid &= (consumed >= 0) & (consumed + length <= CONTEXT)
        if PREFILL:
            valid &= tl.load(Pending + slot) < 0
        else:
            valid &= tl.load(Pending + slot) >= 0
        if not valid:
            tl.store(Status + slot, 3)
            tl.store(Error + slot, 32)


@dataclass(frozen=True)
class DecoderInputs:
    row_slot_ids: torch.Tensor
    row_generations: torch.Tensor
    live_rows: torch.Tensor
    query_lengths: torch.Tensor
    token_ids: torch.Tensor
    prefill_positions: torch.Tensor
    page_tables: torch.Tensor
    sample_enabled: torch.Tensor
    vision_token_indices: torch.Tensor

    @classmethod
    def allocate(cls, device, *, context: int, tokens: int = 512, capacity: int = 4):
        if capacity not in BATCH_CAPACITIES:
            raise ValueError("decoder request capacity must be a power of two through 64")
        if (type(context) is not int or type(tokens) is not int
                or context <= 0 or context % 64 or tokens <= 0):
            raise ValueError("decoder capacities require positive tokens and a 64-aligned context")
        return cls(
            torch.full((capacity,), -1, dtype=torch.int32, device=device),
            torch.zeros(capacity, dtype=torch.int64, device=device),
            torch.zeros(1, dtype=torch.int32, device=device),
            torch.zeros(capacity, dtype=torch.int32, device=device),
            torch.full((tokens,), 248044, dtype=torch.int64, device=device),
            torch.zeros((tokens, 3), dtype=torch.int64, device=device),
            torch.full((capacity, context // 64), -1, dtype=torch.int32, device=device),
            torch.zeros(capacity, dtype=torch.int32, device=device),
            torch.full((tokens,), -1, dtype=torch.int32, device=device),
        )

    def buffers(self):
        return {name: getattr(self, name) for name in self.__dataclass_fields__}


@triton.jit
def _metadata(Slots, Live, Lengths, Pending, Positions, Active, Delta, Tokens, InputRope,
              Cu, ConvSlots, Fresh, StateIndices, Initial, Final, Checkpoint, Offset,
              NumSeqs, NumTokens, RequestIds, QueryPositions, Rope, SequenceLengths,
              SlotMapping, PhysicalPages, PackedPages, IsPrefilling,
              ROWS: tl.constexpr, TOKENS: tl.constexpr, PAGES: tl.constexpr,
              DELTA_STRIDE: tl.constexpr, PREFILL: tl.constexpr, BLOCK: tl.constexpr, SLOTS: tl.constexpr = 4):
    row = tl.program_id(0)
    rows = tl.arange(0, SLOTS)
    live = tl.minimum(tl.load(Live), ROWS)
    active = tl.load(Active + rows, rows < ROWS, other=0) != 0
    lengths = tl.load(Lengths + rows, rows < ROWS, other=0) if PREFILL else tl.full((SLOTS,), 1, tl.int32)
    lengths = tl.where((rows < live) & active, lengths, 0)
    begin = tl.sum(tl.where(rows < row, lengths, 0), 0)
    length = tl.sum(tl.where(rows == row, lengths, 0), 0)
    total = tl.sum(lengths, 0)
    slot = tl.load(Slots + row)
    valid = (row < live) & (length > 0) & (slot >= 0) & (slot < SLOTS)
    position = tl.load(Positions + row)
    model_slot = tl.where(valid, slot + 1, -1)
    end = position + length
    tl.store(Cu + row, begin)
    if row == 0:
        tl.store(Cu + ROWS, total)
        tl.store(NumTokens, total)
        tl.store(NumSeqs, live)
    tl.store(ConvSlots + row, model_slot)
    tl.store(Fresh + row, valid & (position == 0))
    tl.store(StateIndices + row, tl.where(valid, model_slot, 0))
    tl.store(Initial + row, tl.where(valid & (position > 0), model_slot, 0))
    tl.store(Final + row, tl.where(valid, model_slot, 0))
    tl.store(Checkpoint + row, 0)
    tl.store(Offset + row, 0)
    tl.store(SequenceLengths + row, tl.where(valid, end, 0))
    tl.store(IsPrefilling + row, PREFILL & valid)
    if not PREFILL:
        tl.store(Tokens + begin, tl.load(Pending + row), valid)
    delta = tl.load(Delta + (slot + 1) * DELTA_STRIDE, valid, other=0)
    for block in range(tl.cdiv(TOKENS, BLOCK)):
        local = block * BLOCK + tl.arange(0, BLOCK)
        index = begin + local
        mask = valid & (local < length) & (index < TOKENS)
        logical = position + local
        page = tl.load(PhysicalPages + slot * PAGES + logical // 64, mask, other=-1)
        tl.store(RequestIds + index, row, mask)
        tl.store(QueryPositions + index, logical, mask)
        tl.store(SlotMapping + index, page * 64 + logical % 64, mask)
        for axis in tl.static_range(3):
            rope = tl.load(InputRope + index * 3 + axis, mask, other=0) if PREFILL else logical + delta
            tl.store(Rope + index * 3 + axis, rope, mask)
    for block in range(tl.cdiv(PAGES, BLOCK)):
        page_indices = block * BLOCK + tl.arange(0, BLOCK)
        pages = tl.load(PhysicalPages + slot * PAGES + page_indices, valid & (page_indices < PAGES), other=-1)
        tl.store(PackedPages + row * PAGES + page_indices, pages, page_indices < PAGES)


@triton.jit
def _residual_checkpoint(Source, Saved, Rope, SavedRope, Slots, Active, Cu,
                         SAVED_STRIDE: tl.constexpr, ROPE_STRIDE: tl.constexpr, BLOCK: tl.constexpr, SLOTS: tl.constexpr = 4):
    row, tile = tl.program_id(0), tl.program_id(1)
    channel = tile * BLOCK + tl.arange(0, BLOCK)
    slot = tl.load(Slots + row)
    begin, end = tl.load(Cu + row), tl.load(Cu + row + 1)
    active = (tl.load(Active + row) != 0) & (slot >= 0) & (slot < SLOTS) & (end > begin)
    value = tl.load(Source + (end - 1) * 10240 + channel, active & (channel < 10240), other=0)
    tl.store(Saved + (slot + 1) * SAVED_STRIDE + channel, value, active & (channel < 10240))
    if tile == 0:
        axes = tl.arange(0, 4)
        rope = tl.load(Rope + (end - 1) * 3 + axes, active & (axes < 3), other=-1)
        tl.store(SavedRope + (slot + 1) * ROPE_STRIDE + axes, rope, active & (axes < 3))



@triton.jit
def _hidden_checkpoint(Hidden, Saved, Delta, Sample, Slots, Active, Cu, Positions, Rope,
                       SAVED_STRIDE: tl.constexpr, DELTA_STRIDE: tl.constexpr,
                       PREFILL: tl.constexpr, POSITION_BLOCK: tl.constexpr, BLOCK: tl.constexpr, SLOTS: tl.constexpr = 4):
    row = tl.program_id(0)
    channel = tl.arange(0, BLOCK)
    slot = tl.load(Slots + row)
    valid = (tl.load(Active + row) != 0) & (slot >= 0) & (slot < SLOTS)
    begin, end = tl.load(Cu + row), tl.load(Cu + row + 1)
    valid &= end > begin
    value = tl.load(Hidden + (end - 1) * 2560 + channel, valid & (channel < 2560), other=0)
    tl.store(Sample + row * 2560 + channel, value, channel < 2560)
    tl.store(Saved + (slot + 1) * SAVED_STRIDE + channel, value, valid & (channel < 2560))
    if PREFILL:
        position = tl.load(Positions + row)
        old_delta = tl.load(Delta + (slot + 1) * DELTA_STRIDE, valid, other=0)
        old_next = position + old_delta
        indices = tl.arange(0, POSITION_BLOCK)
        rope = tl.load(Rope + indices, valid & (indices // 3 >= begin) & (indices // 3 < end), other=-1)
        next_position = tl.maximum(old_next, tl.max(rope, 0) + 1)
        tl.store(Delta + (slot + 1) * DELTA_STRIDE, next_position - (position + end - begin), valid)


@triton.jit
def _resume_hidden(Saved, Sample, Slots, Epochs, Live, Generation, Status, Error, Consumed, Pending,
                   SAVED_STRIDE: tl.constexpr, CONSUMED_STRIDE: tl.constexpr, BLOCK: tl.constexpr, SLOTS: tl.constexpr = 4):
    row = tl.program_id(0)
    channel = tl.arange(0, BLOCK)
    slot = tl.load(Slots + row)
    valid = (row < tl.load(Live)) & (slot >= 0) & (slot < SLOTS)
    valid &= tl.load(Generation + slot, valid, other=-1) == tl.load(Epochs + row)
    valid &= tl.load(Status + slot, valid, other=0) == 1
    consumed = tl.load(Consumed + slot * CONSUMED_STRIDE, valid, other=0)
    pending = tl.load(Pending + slot, valid, other=-1)
    ready = (consumed > 0) & (pending < 0)
    if valid & ~ready:
        tl.store(Status + slot, 3)
        tl.store(Error + slot, 33)
    value = tl.load(Saved + (slot + 1) * SAVED_STRIDE + channel,
                    valid & ready & (channel < 2560), other=0)
    tl.store(Sample + row * 2560 + channel, value, channel < 2560)


class PreparedDecoderState:
    """Derive model metadata from persistent generation state inside each graph.

    Prefill has one sequence and consumes explicit token/position uploads.
    Decode packs the pending GPU samples from physical request slots. Epoch
    and terminal-state checks precede any model-state mutation. The scheduler
    validates lengths, vocabulary IDs, context bounds and physical page IDs.
    """

    def __init__(self, inputs: DecoderInputs, requests: DeviceRequestState, pool: HybridStatePool,
                 *, tokens: int, rows: int, mode: str):
        if mode not in ("prefill", "decode") or type(rows) is not int or not 1 <= rows <= inputs.row_slot_ids.numel():
            raise ValueError("decoder mode requires prefill/decode and rows within request capacity")
        if (type(tokens) is not int or tokens <= 0
                or any(tokens > tensor.shape[0] for tensor in
                       (inputs.token_ids, inputs.prefill_positions, inputs.vision_token_indices))):
            raise ValueError("decoder token capacity exceeds its input buffers")
        if (mode == "prefill" and rows != 1) or (mode == "decode" and tokens != rows):
            raise ValueError("prefill has one row; decode has one token per row")
        self.inputs, self.requests, self.pool = inputs, requests, pool
        self.tokens, self.rows, self.mode = tokens, rows, mode
        device = inputs.token_ids.device
        self.gdn = DeltaMetadata.allocate(rows, device)
        self.qsa = QSAMetadata.allocate(tokens=tokens, batch=rows,
            context=inputs.page_tables.shape[1] * 64, device=device)
        self.qsa = QSAMetadata(self.qsa.main_block_table, self.qsa.main_block_table,
            self.gdn.convolution_slots, self.qsa.request_ids, self.qsa.query_positions,
            self.qsa.rope_positions, self.qsa.sequence_lengths, self.gdn.query_start_loc,
            self.gdn.accepted_tokens, self.qsa.is_prefilling, self.qsa.slot_mapping)
        self.token_state = PreparedTokenState(requests, inputs.token_ids[:tokens], self.gdn.query_start_loc,
            inputs.row_slot_ids[:rows], inputs.row_generations[:rows], inputs.live_rows, history=requests.ple_history, history_pad_id=248044)
        self.token_state.prepare()
        self._validate = compile_kernel(_validate,
            (inputs.row_slot_ids, inputs.row_generations, inputs.live_rows, inputs.query_lengths,
             requests.generation, requests.status, requests.error_code, requests.pending_token,
             requests.consumed_tokens, rows, tokens, inputs.page_tables.shape[1] * 64,
             requests.consumed_tokens.stride(0), mode == "prefill", requests.capacity), (requests.capacity, 1, 1))
        self.ple = PLEMetadata(inputs.token_ids[:tokens], self.gdn.query_start_loc,
            self.token_state.history, self.gdn.convolution_slots, self.gdn.fresh,
            self.gdn.accepted_tokens, self.gdn.num_seqs, self.gdn.num_tokens)
        g, q, t = self.gdn, self.qsa, self.token_state
        args = (inputs.row_slot_ids, inputs.live_rows, inputs.query_lengths,
            t.pending_tokens, t.positions, t.active, pool.rope_delta, inputs.token_ids, inputs.prefill_positions,
            g.query_start_loc, g.convolution_slots, g.fresh, g.state_indices, g.initial_state_indices,
            g.final_state_indices, g.checkpoint_state_indices, g.checkpoint_offsets, g.num_seqs, g.num_tokens,
            q.request_ids, q.query_positions, q.rope_positions, q.sequence_lengths, q.slot_mapping,
            inputs.page_tables, q.main_block_table, q.is_prefilling,
            rows, tokens, inputs.page_tables.shape[1], pool.rope_delta.stride(0), mode == "prefill", 128, requests.capacity)
        self._metadata = compile_kernel(_metadata, args, (rows, 1, 1))
        self._residual = None

    def prepare_hidden(self, hidden, sample):
        self._hidden = compile_kernel(_hidden_checkpoint,
            (hidden, self.pool.last_hidden, self.pool.rope_delta, sample, self.inputs.row_slot_ids,
             self.token_state.active, self.gdn.query_start_loc, self.token_state.positions, self.qsa.rope_positions,
             self.pool.last_hidden.stride(0), self.pool.rope_delta.stride(0),
             self.mode == "prefill", triton.next_power_of_2(self.tokens * 3), 4096, self.requests.capacity), (self.rows, 1, 1))

    def prepare_residual(self, residual):
        if self.pool.layout.speculative_tokens:
            saved = self.pool.region("target_residual")
            rope = self.pool.region("target_residual_position")
            self._residual = compile_kernel(_residual_checkpoint,
                (residual, saved, self.qsa.rope_positions, rope, self.inputs.row_slot_ids, self.token_state.active,
                 self.gdn.query_start_loc, saved.stride(0), rope.stride(0), 1024, self.requests.capacity), (self.rows, 10, 1))

    def begin(self):
        self._validate()
        self.token_state.gather()
        self.qsa.request_ids.fill_(-1)
        self.qsa.query_positions.fill_(-1)
        self.qsa.rope_positions.fill_(-1)
        self.qsa.slot_mapping.fill_(-1)
        self._metadata()

    def finish(self):
        self._hidden()
        if self._residual is not None:
            self._residual()
        self.token_state.consume()


class PreparedResume:
    """Sample directly from an exact restored prefix's final hidden vector."""

    def __init__(self, inputs, requests, pool, sample_hidden, logits, lm_head_weight,
                 completion_records, frame_index, submission_sequence, *, session, name,
                 local_logits=None, vocabulary_runtime=None, coordinator=None):
        from flare.components.linear import UnquantizedProjection
        from flare.components.sampling import PreparedSampling

        rows = sample_hidden.shape[0]
        if not 1 <= rows <= requests.capacity or logits.shape != (rows, requests.vocab_size):
            raise ValueError("resume hidden/logit row capacities must agree")
        self._gather = compile_kernel(_resume_hidden,
            (pool.last_hidden, sample_hidden, inputs.row_slot_ids, inputs.row_generations, inputs.live_rows,
             requests.generation, requests.status, requests.error_code, requests.consumed_tokens,
             requests.pending_token, pool.last_hidden.stride(0), requests.consumed_tokens.stride(0), 4096, requests.capacity),
            (rows, 1, 1))
        self.sample_hidden = sample_hidden
        if vocabulary_runtime is None:
            self.projection = UnquantizedProjection(sample_hidden, lm_head_weight, logits, session=session, name=name)
        else:
            from flare.components.vocabulary import PreparedVocabularyProjection

            self.projection = PreparedVocabularyProjection(sample_hidden, lm_head_weight, local_logits, logits,
                runtime=vocabulary_runtime, session=session, coordinator=coordinator, name=name)
        self.sampler = PreparedSampling(logits, requests, inputs.row_slot_ids[:rows],
            inputs.row_generations[:rows], inputs.live_rows, completion_records, frame_index, submission_sequence,
            sample_enabled=inputs.sample_enabled[:rows])
        self.sampler.prepare()

    def __call__(self):
        self._gather()
        self.projection(self.sample_hidden)
        self.sampler()
