"""Sliding windows, overlapping C4 retrieval, and dense C128 history."""

from dataclasses import dataclass
from types import SimpleNamespace

import torch
import triton
import triton.language as tl

from ignition.components.attention_output import PreparedGroupedOutputProjection
from ignition.components.compressed_mla import (
    PreparedCompressedCacheWriter,
    PreparedCompressedMLA,
)
from ignition.components.gated_compression import PreparedGatedCompressor
from ignition.components.compressor_overlap import ConcurrentGatedCompression
from ignition.components.linear import QuantizedProjection, UnquantizedProjection
from ignition.components.projection_overlap import ConcurrentProjectionChain
from ignition.components.fp8_indexer import PreparedFP8Indexer
from ignition.components.rms_norm import PreparedRMSNorm
from ignition.components.rotary import PreparedRotaryAngles

from .config import Compressor
from .pool import PAGE_TOKENS, latent_page_bytes

# Select a 4096-token prefill in one call; sequential layers share the planned FP8 workspace.
INDEX_ROWS = 4096
SHARED_ROTARY_ROWS = 32


@triton.jit
def _transfer_window(
    Source,
    Destination,
    StateRows,
    Starts,
    Counts,
    SOURCE_STRIDE: tl.constexpr,
    DEST_STRIDE: tl.constexpr,
    STAGED_TOKENS: tl.constexpr,
    COMMIT: tl.constexpr,
    BLOCK: tl.constexpr,
):
    request, item = tl.program_id(0), tl.program_id(1)
    row = tl.load(StateRows + request)
    offset = tl.arange(0, BLOCK)
    if COMMIT:
        length = tl.load(Counts + request)
        token = tl.maximum(0, length - 128) + item
        live = (row > 0) & (token < length)
        source_row, source_token = request, 128 + token
        dest_row, dest_token = row, (tl.load(Starts + request) + token) % 128
        source_extent: tl.constexpr = STAGED_TOKENS
        dest_extent: tl.constexpr = 128
    else:
        live = row > 0
        source_row, source_token = row, item
        dest_row, dest_token = request, item
        source_extent: tl.constexpr = 128
        dest_extent: tl.constexpr = STAGED_TOKENS
    source_offset = tl.where(
        offset < 576, source_token * 576 + offset, source_extent * 576 + source_token * 8 + offset - 576
    )
    dest_offset = tl.where(
        offset < 576, dest_token * 576 + offset, dest_extent * 576 + dest_token * 8 + offset - 576
    )
    value = tl.load(Source + source_row * SOURCE_STRIDE + source_offset, live & (offset < 584), other=0)
    tl.store(Destination + dest_row * DEST_STRIDE + dest_offset, value, live & (offset < 584))


def transfer_window(source, destination, metadata, staged_tokens, *, counts=None):
    torch.library.wrap_triton(_transfer_window)[(metadata.starts.numel(), 128)](
        source,
        destination,
        metadata.state_rows,
        metadata.starts,
        metadata.lengths if counts is None else counts,
        source.stride(0),
        destination.stride(0),
        staged_tokens,
        counts is not None,
        1024,
    )


@triton.jit
def _hadamard(Values, Output, ROWS: tl.constexpr):
    row = tl.program_id(0)
    dim = tl.arange(0, 128)
    value = tl.load(Values + row * 128 + dim)
    value = value.to(tl.float32)
    for bit in tl.static_range(7):
        other = tl.gather(value, dim ^ (1 << bit), axis=0)
        value = tl.where((dim & (1 << bit)) == 0, value + other, other - value)
    tl.store(Output + row * 128 + dim, value * (128**-0.5))


@triton.jit
def _normalize_rotate(Query, Positions, Frequencies, ROW_STRIDE, HEAD_STRIDE, EPS, WIDTH: tl.constexpr,
                      Angles=None, SHARED_ANGLES: tl.constexpr=False):
    """RMS-normalize one row's head with BF16 rounding at every step, then rotate its last 64 values."""
    row, head = tl.program_id(0), tl.program_id(1)
    base = Query + row * ROW_STRIDE + head * HEAD_STRIDE
    offsets = tl.arange(0, WIDTH)
    value = tl.load(base + offsets).to(tl.float32)
    square = (value * value).to(tl.bfloat16).to(tl.float32)
    mean = (tl.sum(square, axis=0) / WIDTH).to(tl.bfloat16).to(tl.float32)
    scale = tl.rsqrt((mean + EPS).to(tl.bfloat16).to(tl.float32)).to(tl.bfloat16).to(tl.float32)
    tl.store(base + offsets, (value * scale).to(tl.bfloat16), mask=offsets < WIDTH - 64)
    pair = tl.arange(0, 32)
    if SHARED_ANGLES:
        cos = tl.load(Angles + row * 64 + 2 * pair)
        sin = tl.load(Angles + row * 64 + 2 * pair + 1)
    else:
        angle = tl.load(Positions + row).to(tl.float32) * tl.load(Frequencies + pair)
        cos, sin = tl.cos(angle), tl.sin(angle)
    tail = base + WIDTH - 64 + 2 * pair
    a = (tl.load(tail).to(tl.float32) * scale).to(tl.bfloat16).to(tl.float32)
    b = (tl.load(tail + 1).to(tl.float32) * scale).to(tl.bfloat16).to(tl.float32)
    tl.store(tail, (a * cos - b * sin).to(tl.bfloat16))
    tl.store(tail + 1, (a * sin + b * cos).to(tl.bfloat16))


def normalize_rotate(query, positions, frequencies, eps, *, angles=None):
    """Normalize and rotate (rows, heads, width) BF16 queries in place in one launch."""
    torch.library.wrap_triton(_normalize_rotate)[(query.shape[0], query.shape[1])](
        query, positions, frequencies, query.stride(0), query.stride(1), eps, query.shape[2],
        angles, angles is not None
    )


def rotate_index(source, destination):
    torch.library.wrap_triton(_hadamard)[(source.numel() // 128,)](source, destination, source.numel() // 128)


@dataclass(frozen=True)
class AttentionScratch:
    projected: torch.Tensor
    query_low: torch.Tensor
    query: torch.Tensor
    window_values: torch.Tensor
    attended: torch.Tensor
    output: torch.Tensor
    selected: torch.Tensor
    dense_indices: torch.Tensor
    index_query: torch.Tensor
    index_weights: torch.Tensor

    @classmethod
    def allocate(cls, tokens, world, device, *, blocks=8192):
        def zeros(shape, dtype=torch.bfloat16):
            return torch.zeros(shape, dtype=dtype, device=device)

        return cls(
            zeros((tokens, 1536)),
            zeros((tokens, 1024)),
            zeros((tokens, 64 // world, 512)),
            zeros((tokens, 512)),
            zeros((tokens, 64 // world, 512)),
            zeros((tokens, 4096)),
            zeros((tokens, 512), torch.int32) - 1,
            zeros((tokens, blocks * (PAGE_TOKENS // 128)), torch.int32) - 1,
            zeros((tokens, 64, 128)),
            zeros((tokens, 64)),
        )


class AttentionIndices:
    """Window and compression indices that every attention layer of one forward shares.

    They depend only on the batch metadata, so the forward computes them once before its first layer
    instead of once per layer.
    """

    def __init__(self, metadata, *, mode, tokens, visible, buffers):
        rows = metadata.starts.numel()
        self.metadata, self.mode, self.visible, self.buffers = metadata, mode, visible, buffers
        self.window_page_size = 128 if mode == "decode" else 128 + tokens // rows
        width = self.window_page_size if mode == "draft" else 128 + (384 if visible is not None else 0)
        self.offsets = torch.arange(width, device=metadata.positions.device)
        self.window_slots = buffers.full((tokens,), -1, torch.int64)
        self.window_indices = buffers.full((tokens, width), -1, torch.int32)
        self.window_lengths = buffers.zeros((tokens,), torch.int32)
        self.ratios, self.dense = {}, None
        self.angles = {}

    def rotary_angles(self, rotary, *, compressed, ratio=0):
        key = compressed, ratio
        if key not in self.angles:
            positions = self.metadata.positions if ratio == 0 else self.compressed(ratio).position
            self.angles[key] = PreparedRotaryAngles(positions, rotary.frequencies[compressed], buffers=self.buffers)
        return self.angles[key]

    def compressed(self, ratio):
        """Slots, lengths and emission flags shared by the layers compressing at `ratio`."""
        if ratio not in self.ratios:
            tokens, buffers = self.window_slots.shape[0], self.buffers
            self.ratios[ratio] = SimpleNamespace(
                slots=buffers.full((tokens,), -1, torch.int64),
                lengths=buffers.zeros((tokens,), torch.int32),
                index_lengths=buffers.zeros((tokens,), torch.int32),
                active_width=buffers.zeros((1,), torch.int32),
                position=buffers.zeros((tokens,), torch.int64),
                emitted=buffers.zeros((tokens,), torch.bool),
                emitted_slots=buffers.full((tokens,), -1, torch.int64),
            )
        return self.ratios[ratio]

    def dense_pages(self, target):
        """Dense C128 layers read one copy of the token page table."""
        if self.dense is None:
            self.dense = target
        elif self.dense.data_ptr() != target.data_ptr() or self.dense.shape != target.shape:
            raise ValueError("dense C128 layers of one forward share their page indices")
        return self.dense

    def __call__(self):
        m = self.metadata
        live, position = m.positions >= 0, m.positions.clamp_min(0)
        rows, starts = m.state_rows[m.requests], m.starts[m.requests]
        left = (position - 127).clamp_min(0)
        right = position
        if self.visible is not None:
            image_left, image_right = self.visible
            left = torch.minimum(left, position - image_left)
            right = position + image_right
        self.window_lengths.copy_(torch.where(live, right - left + 1, 0))
        positions = left[:, None] + self.offsets
        if self.mode == "decode":
            slots = rows * 128 + position % 128
            indices = rows[:, None] * 128 + positions % 128
        else:
            base = m.requests * self.window_page_size
            slots = base + 128 + position - starts
            indices = torch.where(
                positions < starts[:, None], positions % 128, 128 + positions - starts[:, None]
            )
            indices = indices + base[:, None]
            if self.mode == "draft":
                context = starts.clamp(max=128)
                self.window_lengths.copy_(torch.where(live, context + m.lengths[m.requests], 0))
                indices = base[:, None] + torch.where(
                    self.offsets < context[:, None], self.offsets, 128 + self.offsets - context[:, None]
                )
        self.window_slots.copy_(torch.where(live, slots, -1))
        self.window_indices.copy_(torch.where(self.offsets < self.window_lengths[:, None], indices, -1))
        for ratio, shared in self.ratios.items():
            shared.lengths.copy_(torch.where(live, (position + 1) // ratio, 0))
            shared.index_lengths.copy_(shared.lengths.clamp(max=512) if ratio == 4 else shared.lengths)
            shared.active_width.copy_(shared.lengths.amax().view(1))
            pages = m.page_table[m.requests, (position // PAGE_TOKENS).clamp(max=m.page_table.shape[1] - 1)]
            shared.slots.copy_(
                torch.where(live, pages.long() * (PAGE_TOKENS // ratio) + (position % PAGE_TOKENS) // ratio, -1)
            )
            shared.position.copy_(((position + 1) // ratio * ratio - ratio).clamp_min(0))
            shared.emitted.copy_(live & ((m.positions + 1) % ratio == 0))
            shared.emitted_slots.copy_(torch.where(shared.emitted, shared.slots, -1))
        if self.dense is not None:
            records = PAGE_TOKENS // 128
            offsets = torch.arange(records, device=m.page_table.device)
            pages = m.token_page_table[:, :, None]
            self.dense.copy_(torch.where(pages >= 0, pages * records + offsets, -1).flatten(1))
        for angles in self.angles.values():
            angles()


class PreparedAttention:
    def __init__(
        self,
        layer,
        weights,
        source,
        scratch,
        metadata,
        pool,
        rotary,
        *,
        mode,
        world,
        session,
        name,
        workspace,
        buffers,
        visible=None,
        indices=None,
    ):
        self.scratch, self.metadata, self.pool = scratch, metadata, pool
        self.rotary, self.mode, self.visible = rotary, mode, visible
        self.ratio = 0 if mode == "draft" else pool.config.text["compress_ratios"][layer]
        self.eps = pool.config.text["rms_norm_eps"]
        tokens, rows, device = source.shape[0], metadata.starts.numel(), source.device
        self.window = pool.window(layer, draft=mode == "draft")
        # Layers of one forward share their window and compression indices; a standalone layer owns them.
        self.owns_indices = indices is None
        if indices is None:
            indices = AttentionIndices(metadata, mode=mode, tokens=tokens, visible=visible, buffers=buffers)
        self.indices = indices
        self.window_page_size = indices.window_page_size
        # Every launch stages its window and rewrites it before reading it, so it lives in the graph's
        # private buffers, which graphs that never run concurrently share.
        self.window_cache = (
            self.window
            if mode == "decode"
            else buffers.zeros((rows, latent_page_bytes(self.window_page_size)), torch.uint8)
        )
        self.window_slots, self.window_indices = indices.window_slots, indices.window_indices
        self.window_lengths = indices.window_lengths
        shared = indices.compressed(self.ratio) if self.ratio else None
        # Small graphs share repeated trigonometry; large prefills avoid table-read traffic.
        share_angles = tokens <= SHARED_ROTARY_ROWS
        self.angles = indices.rotary_angles(rotary, compressed=bool(self.ratio)) if share_angles else None
        self.compressed_angles = (indices.rotary_angles(rotary, compressed=True, ratio=self.ratio)
                                  if self.ratio and share_angles else None)
        if shared is not None:
            self.compressed_slots, self.lengths = shared.slots, shared.lengths
            self.active_width, self.index_lengths = shared.active_width, shared.index_lengths
        self.overlap_compression = self.ratio == 4 and mode == "verify" and tokens <= 8
        # Small C4 verification overlaps compression. Quantized projections leave room for that
        # branch; tuning the projections in isolation does not measure their combined latency.
        activation_mode = "quantized" if self.overlap_compression else "auto"
        self.qkv = QuantizedProjection(
            source,
            weights["qkv_a"],
            scratch.projected,
            session=session,
            name=name + ".qkv_a",
            workspace=workspace,
            activation_mode=activation_mode,
        )
        self.qb = QuantizedProjection(
            scratch.query_low,
            weights["q_b"],
            scratch.query.view(tokens, -1),
            session=session,
            name=name + ".q_b",
            workspace=workspace,
            activation_mode=activation_mode,
        )
        self.qnorm = PreparedRMSNorm(
            scratch.projected[:, :1024],
            weights["q_norm"],
            scratch.query_low,
            eps=self.eps,
            session=session,
            name=name + ".q_norm",
        )
        self.kvnorm = PreparedRMSNorm(
            scratch.projected[:, 1024:],
            weights["kv_norm"],
            scratch.window_values,
            eps=self.eps,
            session=session,
            name=name + ".kv_norm",
        )
        self.output_projection = PreparedGroupedOutputProjection(
            scratch.attended.view(tokens, 8 // world, 4096),
            weights["output"],
            scratch.output,
            session=session,
            name=name + ".output",
            workspace=workspace,
        )

        def writer(values, cache, slots, size, kind, suffix):
            return PreparedCompressedCacheWriter(
                values,
                cache,
                slots,
                page_size=size,
                kind=kind,
                session=session,
                name=name + suffix,
                cache_format="deepseek_v4",
            )

        self.window_writer = writer(
            scratch.window_values,
            self.window_cache,
            self.window_slots,
            self.window_page_size,
            "swa",
            ".window_write",
        )
        self.compressor = self.index_compressor = None
        self.selectors = []
        self.indexed_indices = None
        indexed = {}
        self.compression = None
        if self.ratio:
            specs = [Compressor(layer, self.ratio, 512)]
            if self.ratio == 4:
                specs.append(Compressor(layer, 4, 128, index=True))
            widths = [spec.state_shape[1] for spec in specs]
            self.compressed = buffers.zeros((tokens, 2 * sum(widths)), torch.float32)
            self.compression = UnquantizedProjection(
                source, weights["compression"], self.compressed, session=session, name=name + ".compression"
            )
            projected, offset = [], 0
            for width in widths:
                projected.append((self.compressed[:, offset:offset + width],
                                  self.compressed[:, offset + width:offset + 2 * width]))
                offset += 2 * width
            self.compressor = PreparedGatedCompressor(
                specs[0],
                weights["compressor"],
                source,
                metadata,
                pool,
                self.compressed_slots,
                session=session,
                name=name + ".compress",
                deferred=mode == "verify",
                buffers=buffers,
                emitted=shared.emitted,
                emitted_slots=shared.emitted_slots,
                projected=projected[0],
            )
            self.compressed_writer = writer(
                self.compressor.output,
                pool.pages[f"layer{layer}.latent_page"],
                self.compressor.slots,
                PAGE_TOKENS // self.ratio,
                "indexed",
                ".compressed_write",
            )
            if self.ratio == 4:
                self.index_compressor = PreparedGatedCompressor(
                    specs[1],
                    weights["index_compressor"],
                    source,
                    metadata,
                    pool,
                    self.compressed_slots,
                    session=session,
                    name=name + ".index_compress",
                    deferred=mode == "verify",
                    buffers=buffers,
                    emitted=shared.emitted,
                    emitted_slots=shared.emitted_slots,
                    projected=projected[1],
                )
                self.index_query = QuantizedProjection(
                    scratch.query_low,
                    weights["index_q"],
                    scratch.index_query.view(tokens, -1),
                    session=session,
                    name=name + ".index_query",
                    workspace=workspace,
                    activation_mode=activation_mode,
                )
                self.index_weight = UnquantizedProjection(
                    source,
                    weights["index_weights"],
                    scratch.index_weights,
                    session=session,
                    name=name + ".index_weight",
                )
                for first in range(0, tokens, INDEX_ROWS):
                    last = min(tokens, first + INDEX_ROWS)
                    self.selectors.append(
                        PreparedFP8Indexer(
                            scratch.index_query[first:last],
                            scratch.index_weights[first:last],
                            pool.pages[f"layer{layer}.index_page"],
                            metadata.token_page_table[first:last],
                            self.lengths[first:last],
                            self.active_width,
                            scratch.selected[first:last],
                            mode="prefill" if mode == "prefill" else "decode",
                            session=session,
                            name=f"{name}.index{first}",
                            workspace=workspace,
                            buffers=buffers,
                            prefill_records_per_call=131072,
                        )
                    )
                selected = scratch.selected
            else:
                selected = indices.dense_pages(scratch.dense_indices[:,
                    : metadata.page_table.shape[1] * (PAGE_TOKENS // 128)])
            self.indexed_indices = selected
            indexed = {
                "indexed": pool.pages[f"layer{layer}.latent_page"],
                "indexed_indices": selected,
                "indexed_lengths": self.index_lengths,
                "indexed_page_size": PAGE_TOKENS // self.ratio,
            }
        self.concurrent_compression = None
        self.concurrent_index = None
        if self.overlap_compression:
            self.concurrent_index = ConcurrentProjectionChain(
                (self.index_query,), parallel_plans=(self.kvnorm.plan, self.qb.plan, self.window_writer.plan),
                session=session, workspace=workspace)
            compressors = [self.compressor]
            if self.index_compressor is not None:
                compressors.append(self.index_compressor)
            self.concurrent_compression = ConcurrentGatedCompression(
                self.compression, compressors, self.compressed_writer,
                indexer=self.selectors[0] if self.selectors else None,
                index_weight=self.index_weight if self.selectors else None,
                metadata=metadata, positions=self.indices.compressed(self.ratio).position,
                frequencies=self.rotary.frequencies[True],
                angles=self.compressed_angles.values,
                parallel_plans=(self.qkv.plan, self.qnorm.plan, self.kvnorm.plan,
                                self.qb.plan, self.window_writer.plan),
                session=session, workspace=workspace)
        self.attention = PreparedCompressedMLA(
            scratch.query,
            self.window_cache,
            self.window_indices,
            self.window_lengths,
            weights["sink"],
            scratch.attended,
            window_page_size=self.window_page_size,
            # Draft indices expose the complete noncausal block. The decode kernel follows
            # those explicit lists; its tiling does not impose a causal attention mask.
            mode="decode" if mode in ("decode", "verify") or mode == "draft" and tokens <= 32 else "extend",
            session=session,
            name=name + ".attention",
            workspace=workspace,
            cache_format="deepseek_v4",
            **indexed,
        )

    def rotate(self, tensor, *, compression=False, inverse=False):
        angles = self.compressed_angles if compression else self.angles
        if angles is not None:
            angles.apply(tensor, inverse=inverse)
        else:
            position = (self.indices.compressed(self.ratio).position if compression
                        else self.metadata.positions.clamp_min(0))
            self.rotary.apply(tensor, position, compressed=compression or bool(self.ratio), inverse=inverse)

    def __call__(self, source):
        m, s = self.metadata, self.scratch
        position = m.positions if self.angles is not None else m.positions.clamp_min(0)
        if self.owns_indices:
            self.indices()
        if self.mode != "decode":
            transfer_window(self.window, self.window_cache, m, self.window_page_size)
        if self.concurrent_compression is not None:
            self.concurrent_compression.start(source)
        self.qkv(source)
        self.qnorm(s.projected[:, :1024])
        if self.concurrent_index is not None:
            self.concurrent_index.start(s.query_low)
        self.kvnorm(s.projected[:, 1024:])
        self.qb(s.query_low)
        normalize_rotate(s.query, position, self.rotary.frequencies[bool(self.ratio)], self.eps,
                         angles=None if self.angles is None else self.angles.values)
        self.rotate(s.window_values)
        self.window_writer(s.window_values)
        if self.concurrent_compression is not None:
            self.concurrent_index.finish(s.query_low, s.query)
            self.rotate(s.index_query)
            self.concurrent_compression.finish(source, [s.query, self.window_cache, s.index_query])
        if self.ratio:
            if self.concurrent_compression is None:
                self.compression(source)
                latent = self.compressor(source)
                self.rotate(latent, compression=True)
                self.compressed_writer(latent)
            if self.index_compressor is not None:
                if self.concurrent_compression is None:
                    keys = self.index_compressor(source)
                    self.rotate(keys, compression=True)
                    self.selectors[0].write_keys(keys, self.index_compressor.slots)
                if self.concurrent_compression is None:
                    self.index_query(s.query_low)
                    self.rotate(s.index_query)
                    self.index_weight(source)
                for selector in self.selectors:
                    selector()
                logical = s.selected.long()
                records = PAGE_TOKENS // 4
                physical = m.token_page_table.gather(1, (logical // records).clamp(0, m.page_table.shape[1] - 1))
                s.selected.copy_(
                    torch.where((logical >= 0) & (physical >= 0), physical * records + logical % records, -1)
                )
        self.attention()
        self.rotate(s.attended, inverse=True)
        self.output_projection(s.attended.view(s.attended.shape[0], -1, 4096))
        if self.mode == "prefill":
            self.commit(m.lengths)
        return s.output

    def commit(self, counts):
        transfer_window(self.window_cache, self.window, self.metadata, self.window_page_size, counts=counts)
        if self.mode == "verify":
            if self.compressor is not None:
                self.compressor.commit(counts)
            if self.index_compressor is not None:
                self.index_compressor.commit(counts)
